File: src/Biweekly_Contest_21/Find_the_Longest_Substring_Containing_Vowels_in_Even_Counts/Solution.py
class Solution:
    def findTheLongestSubstring(self, s: str) -> int:
        vowels = ('a', 'e', 'i', 'o', 'u')
        map = {}
        state = [True, True, True, True, True]
        map[tuple(state)] = -1
        maxLength = 0

        for index in range(0, len(s)):
            c = s[index]
            for i in range(5):
                if c == vowels[i]:
                    state[i] = not state[i]
                    break

            if tuple(state) not in map:
                map[tuple(state)] = index
            else:
                maxLength = max(maxLength, index - map[tuple(state)])

        return maxLength

File: src/Biweekly_Contest_21/Find_the_Longest_Substring_Containing_Vowels_in_Even_Counts/Solution.kt
package Biweekly_Contest_21.b

import kotlin.math.max

fun main() {
    Solution().findTheLongestSubstring("leetcodeisgreat")
}

//https://leetcode.com/contest/biweekly-contest-21/problems/find-the-longest-substring-containing-vowels-in-even-counts/
class Solution {
    //false:single true:double
    data class State(
        val a: Boolean,
        val e: Boolean,
        val i: Boolean,
        val o: Boolean,
        val u: Boolean
    )

    fun findTheLongestSubstring(s: String): Int {
        val map = hashMapOf<State, Int>()
        var state = State(true, true, true, true, true)
        map[state] = -1
        var maxLength = 0

        s.forEachIndexed { index, c ->
            state = when (c) {
                'a' -> state.copy(a = !state.a)
                'e' -> state.copy(e = !state.e)
                'i' -> state.copy(i = !state.i)
                'o' -> state.copy(o = !state.o)
                'u' -> state.copy(u = !state.u)
                else -> state
            }

            if (state !in map) {
                map[state] = index
            } else {
                maxLength = max(maxLength, index - map[state]!!)
            }
        }

        return maxLength

    }
}